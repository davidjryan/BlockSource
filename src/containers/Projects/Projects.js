import React, { Component } from 'react';
import {Card, CardActions, CardHeader, CardTitle, CardText} from 'material-ui/Card';
import './Projects.css';
import { connect } from 'react-redux';
import { projectsEndpoint, contributorsEndpoint } from '../../actions';
import FlatButton from 'material-ui/FlatButton';
import Dialog from 'material-ui/Dialog';

export class Projects extends Component {
  state = {
    open: false,
  };

  handleOpen = () => {
    this.setState({open: true});
  };

  handleClose = () => {
    this.setState({open: false});
  };

  componentDidMount() {
    this.props.handleProjectsFetch();
  }

  checkContributors() {
    if (!this.props.contributors) {
      return <p>Nobody has backed this project</p>
    } else {
    this.props.contributors.map(contributor => <p>{contributor.name}</p>)
    }
  }

  mapProjects() {
    const actions = [
      <FlatButton
        label="Cancel"
        primary={true}
        onClick={this.handleClose}
      />,
      <FlatButton
        label="Submit"
        primary={true}
        keyboardFocused={true}
        onClick={this.handleClose}
      />,
    ];

    return this.props.projects.map( (project, index) => {
      return <Card key={project.id}>
        <CardHeader
          title={project.title}
          subtitle={`Goal Amount: ${project.goal_amount} Fund Amount: ${project.fund_amount}`}
        />
        <CardText>
          {project.description}
        </CardText>
        <CardActions actAsExpander={true}>
          <FlatButton label="Contributors" primary={true} onClick={() => this.props.showContributors(project.id)}/>
        </CardActions>
        <CardText expandable={true}>
        
          {!!this.props.contributors.length ? this.props.contributors.map(contributor => <p>{contributor.name}</p>) : <p>Nobody has backed this project</p>}    
        </CardText>
        <FlatButton label="Add Contribution" onClick={this.handleOpen} />
        <Dialog
          title={`Contribute to ${project.title}`}
          actions={actions}
          modal={false}
          open={this.state.open}
          onRequestClose={this.handleClose}
        >
        </Dialog>
      </Card>;
    });
  }

  render() {

    return (
      <div className='projects-component'>
        {this.props.projects && this.mapProjects()}
      </div>
    );
  }
}

export const mapStateToProps = store => ({
  projects: store.projects,
  contributors: store.projectContributors
});

export const mapDispatchToProps = (dispatch) => {
  return {
    handleProjectsFetch: () => {
      dispatch(projectsEndpoint());
    },
    showContributors: (id) => {
      dispatch(contributorsEndpoint(id));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Projects);