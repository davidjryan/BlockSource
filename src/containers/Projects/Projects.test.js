import React from 'react';
import { Projects, mapStateToProps, mapDispatchToProps } from './Projects';
import { shallow } from 'enzyme';

describe('Projects Container', () => {
  describe('Projects', () => {
    let projects;
    const mockHandleProjectsFetch = jest.fn();
    const mockShowContributors = jest.fn();

    beforeEach( () => {
      projects = shallow(
        <Projects 
          handleProjectsFetch={mockHandleProjectsFetch}
          showContributors={mockShowContributors}
        />);
    });

    it('should exist', () => {
      expect(projects).toBeDefined();
    });

    it('should match the snapshot', () => {
      expect(projects).toMatchSnapshot();
    });

    it('should call loginUser on form submit', () => {
      projects.instance().componentDidMount();
      expect(mockHandleProjectsFetch).toHaveBeenCalled();
    });

  });

  describe('mapStateToProps', () => {
    let mockStore;
    let result;

    beforeAll(() => {
      mockStore = {
        projects: ['project', 'project2']
      }
      result = mapStateToProps(mockStore);
    });

    it('should pull projects from the store', () => {
      expect(result.projects).toEqual(mockStore.projects);
    });

  });

  describe('mapDispatchToProps', () => {
    let mockDispatch;
    let result;

    beforeEach(() => {
      mockDispatch = jest.fn();
      result = mapDispatchToProps(mockDispatch);
    });

    it('should call dispatch when handleProjectsFetch is called', () => {

      result.handleProjectsFetch();
      expect(mockDispatch).toHaveBeenCalled();
    });

    it('should call dispatch when showContributors is called', () => {
      const mockProjectId = 'id';

      result.showContributors(mockProjectId);
      expect(mockDispatch).toHaveBeenCalled();
    });

  });
});