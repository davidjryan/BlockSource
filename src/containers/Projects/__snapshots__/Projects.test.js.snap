// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects Container Projects should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Projects
    handleProjectsFetch={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      }
    }
    showContributors={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": undefined,
      "className": "projects-component",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "className": "projects-component",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
